src/Features/Core/Portable/RemoveAsyncModifier/AbstractRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,236 @@
+                    // body then we've done all we can
+                    if (blockBodiedNode != null)
+                    {
+                        editor.ReplaceNode(replacementNode, blockBodiedNode);
+                        // We need to get the block inside the block bodied method to know where to add the return
+                        var node = GetLastChildOfBlock(blockBodiedNode);
+                        AppendTaskCompletedTaskReturn(node, editor, knownTypes);
+                    }
+                }
+                else
+                {
+                    // For Task<T> returning expression bodied methods we can just wrap the whole expression
+                    WrapExpressionWithTaskFromResult(expressionBody, editor, knownTypes);
+                }
+            }
+            else
+            {
+                // Block bodied methods might have return statements so update them to task returning
+                ChangeReturnStatements(replacementNode, editor, knownTypes);
+
+                // An "async Task" method has an implicit return at the end, so removing async means
+                // we need to insert it explicitly if the end of the method is reachable.
+                if (methodSymbol.ReturnType == knownTypes._taskType)
+                {
+                    // We have to use the original node to do control flow analysis, but the reachability of it is the same
+                    var controlFlow = AnalyzeControlFlow(semanticModel, originalNode);
+
+                    if (controlFlow != null)
+                    {
+                        // For local functions and block bodied lambdas the EndPointIsReachable is false but we still might need to
+                        // insert a return. We can tell by checking for the presence of exit points that aren't return statements
+                        var hasNonReturnExitPoints = controlFlow.ExitPoints.Any<SyntaxNode>(e => !(e is TReturnStatementSyntax));
+                        if (controlFlow.EndPointIsReachable || hasNonReturnExitPoints)
+                        {
+                            var node = GetLastChildOfBlock(replacementNode);
+                            AppendTaskCompletedTaskReturn(node, editor, knownTypes);
+                        }
+                    }
+                }
+            }
+        }
+
+        private void ChangeReturnStatements(SyntaxNode node, SyntaxEditor editor, KnownTypes knownTypes)
+        {
+            var generator = editor.Generator;
+
+            var returns = node.DescendantNodesAndSelf().Where(n => n is TReturnStatementSyntax);
+            foreach (TReturnStatementSyntax returnSyntax in returns)
+            {
+                // Make sure we're not changing returns in nested local functions or lambdas
+                var containingMethod = returnSyntax.FirstAncestorOrSelf<SyntaxNode>(IsAsyncSupportingFunctionSyntax);
```

Comment:
you can pass a helper to DescendantNodesAndSelf to say when to *stop/continue* recursing.  so isntead of needing this check here, you just ensure you don't dive into nested functions to begin with.  then you avoid the walk down + check up logic.